src/dbnode/storage/index_insert_queue.go

```diff
@@ -259,25 +324,97 @@ func (b *nsIndexInsertBatch) allocateAllInserts() {
 
 func (b *nsIndexInsertBatch) AllInserts() *index.WriteBatch {
 	b.allInserts.Reset()
-	for _, shardInserts := range b.shardInserts {
-		b.allInserts.AppendAll(shardInserts)
+	for _, inserts := range b.insertsByCPUCore {
+		inserts.Lock()
+		for _, shardInserts := range inserts.shardInserts {
+			b.allInserts.AppendAll(shardInserts)
+		}
+		for _, insert := range inserts.batchInserts {
+			b.allInserts.Append(insert.Entry, insert.Document)
+		}
+		inserts.Unlock()
 	}
 	return b.allInserts
 }
 
-func (b *nsIndexInsertBatch) Reset() {
+func (b *nsIndexInsertBatch) Rotate(target *nsIndexInsertBatch) *sync.WaitGroup {
```

Comment:
I don't quite get this rotation optimization, it seems like we rotate into a target batch and then immediately call `AllInserts()` and then `indexBatchFn()`.

Reply:
It's really just for copying one into the other.